### PR TITLE
xcowsay: Init at 1.5

### PR DESCRIPTION
**Pull request by dasJ**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dasJ** [comment]
@GrahamcOfBorg build xcowsay

**Review by aanderse — COMMENTED** [review]
We are trying to eliminate `gtk2` I'm pretty sure. Is there a `gtk3` option?

cc @jtojnar

**Comment by jtojnar** [comment]
It appears the code was made at least partially compatible:

* https://github.com/nickg/xcowsay/commit/4c815256abbceb704e3dbe363defc81f68167e71
* https://github.com/nickg/xcowsay/commit/e8aaf61cab5c6e2b5aaa219a5b692ecc1564f5db
* https://github.com/nickg/xcowsay/commit/b248c779680a810a22987759e295a67b414841a2

But the configure script has GTK2 hardcoded:

https://github.com/nickg/xcowsay/blob/ef6944a0ff7fe8271f546f857454d45d90e53725/configure.ac#L34

Would you mind trying to replace it with `gtk+-3.0` and adding `autoreconfHook` to `nativeBuildInputs`?

**Comment by dasJ** [comment]
Hmm seems like this doesn't work:
```
In file included from display_cow.c:37:
floating_shape.h:30:4: error: unknown type name 'GdkDrawable'
   30 |    GdkDrawable *mask_bitmap;
      |    ^~~~~~~~~~~
display_cow.c: In function 'display_cow':
display_cow.c:244:4: warning: 'gdk_screen_get_n_monitors' is deprecated: Use 'gdk_display_get_n_monitors' instead [-Wdeprecated-declarations]
  244 |    gint n_monitors = gdk_screen_get_n_monitors(screen);
      |    ^~~~
In file included from /nix/store/79k23ba41sqbr95x32zyb0w0k3smkkwk-gtk+3-3.24.12-dev/include/gtk-3.0/gdk/gdkapplaunchcontext.h:31,
                 from /nix/store/79k23ba41sqbr95x32zyb0w0k3smkkwk-gtk+3-3.24.12-dev/include/gtk-3.0/gdk/gdk.h:32,
                 from /nix/store/79k23ba41sqbr95x32zyb0w0k3smkkwk-gtk+3-3.24.12-dev/include/gtk-3.0/gtk/gtk.h:30,
                 from display_cow.c:29:
/nix/store/79k23ba41sqbr95x32zyb0w0k3smkkwk-gtk+3-3.24.12-dev/include/gtk-3.0/gdk/gdkscreen.h:74:14: note: declared here
   74 | gint         gdk_screen_get_n_monitors        (GdkScreen    *screen);
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~
display_cow.c:251:4: warning: 'gdk_screen_get_monitor_geometry' is deprecated: Use 'gdk_monitor_get_geometry' instead [-Wdeprecated-declarations]
  251 |    gdk_screen_get_monitor_geometry(screen, pick, &geom);
      |    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /nix/store/79k23ba41sqbr95x32zyb0w0k3smkkwk-gtk+3-3.24.12-dev/include/gtk-3.0/gdk/gdkapplaunchcontext.h:31,
                 from /nix/store/79k23ba41sqbr95x32zyb0w0k3smkkwk-gtk+3-3.24.12-dev/include/gtk-3.0/gdk/gdk.h:32,
                 from /nix/store/79k23ba41sqbr95x32zyb0w0k3smkkwk-gtk+3-3.24.12-dev/include/gtk-3.0/gtk/gtk.h:30,
                 from display_cow.c:29:
/nix/store/79k23ba41sqbr95x32zyb0w0k3smkkwk-gtk+3-3.24.12-dev/include/gtk-3.0/gdk/gdkscreen.h:78:14: note: declared here
   78 | void         gdk_screen_get_monitor_geometry  (GdkScreen    *screen,
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
display_cow.c: In function 'try_dbus':
display_cow.c:359:4: warning: 'g_type_init' is deprecated [-Wdeprecated-declarations]
  359 |    g_type_init();
      |    ^~~~~~~~~~~
In file included from /nix/store/20jfnqdap9kni2gir3lxb12m4mhrfkx6-glib-2.62.4-dev/include/glib-2.0/gobject/gobject.h:24,
                 from /nix/store/20jfnqdap9kni2gir3lxb12m4mhrfkx6-glib-2.62.4-dev/include/glib-2.0/gobject/gbinding.h:29,
                 from /nix/store/20jfnqdap9kni2gir3lxb12m4mhrfkx6-glib-2.62.4-dev/include/glib-2.0/glib-object.h:23,
                 from /nix/store/20jfnqdap9kni2gir3lxb12m4mhrfkx6-glib-2.62.4-dev/include/glib-2.0/gio/gioenums.h:28,
                 from /nix/store/20jfnqdap9kni2gir3lxb12m4mhrfkx6-glib-2.62.4-dev/include/glib-2.0/gio/giotypes.h:28,
                 from /nix/store/20jfnqdap9kni2gir3lxb12m4mhrfkx6-glib-2.62.4-dev/include/glib-2.0/gio/gio.h:26,
                 from /nix/store/79k23ba41sqbr95x32zyb0w0k3smkkwk-gtk+3-3.24.12-dev/include/gtk-3.0/gdk/gdkapplaunchcontext.h:28,
                 from /nix/store/79k23ba41sqbr95x32zyb0w0k3smkkwk-gtk+3-3.24.12-dev/include/gtk-3.0/gdk/gdk.h:32,
                 from /nix/store/79k23ba41sqbr95x32zyb0w0k3smkkwk-gtk+3-3.24.12-dev/include/gtk-3.0/gtk/gtk.h:30,
                 from display_cow.c:29:
/nix/store/20jfnqdap9kni2gir3lxb12m4mhrfkx6-glib-2.62.4-dev/include/glib-2.0/gobject/gtype.h:691:23: note: declared here
  691 | void                  g_type_init                    (void);
      |                       ^~~~~~~~~~~
display_cow.c: In function 'load_cow':
display_cow.c:91:7: warning: ignoring return value of 'asprintf', declared with attribute warn_unused_result [-Wunused-result]
   91 |       asprintf(&cow_path, "%s/%s_%s.png", DATADIR,
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   92 |          get_string_option("image_base"),
      |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   93 |          get_string_option("cow_size"));
      |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [Makefile:477: display_cow.o] Error 1
make[2]: Leaving directory '/build/xcowsay-1.4/src'
make[1]: *** [Makefile:494: all-recursive] Error 1
make[1]: Leaving directory '/build/xcowsay-1.4'
make: *** [Makefile:371: all] Error 2
builder for '/nix/store/m1sc2cwwmd4lkp62if9p8cf5a4qq59zf-xcowsay-1.4.drv' failed with exit code 2
```

**Comment by jtojnar** [comment]
Thanks for checking. [`GdkDrawable`](https://developer.gnome.org/gdk2/2.24/gdk2-Drawing-Primitives.html) has been [9 years ago](https://gitlab.gnome.org/GNOME/gtk/commit/3d506df2344e7499fbe3946683647ea53f871b1b), cairo should probably be used instead.

Using GTK 2 is fine for now, opening an issue upstream about porting it to GTK 3/cairo/pango would be nice.

**Comment by dasJ** [comment]
I opened [#6](https://github.com/nickg/xcowsay/issues/6)

**Comment by jtojnar** [comment]
```ShellSession
$ result/bin/xcowsay "hullo"
...works fine
$ result/bin/xcowfortune 
env: ‘fortune’: No such file or directory
$ result/bin/xcowdream 
result/bin/xcowdream: line 2: exec: xcowsay: not found
$ result/bin/xcowthink 
result/bin/xcowthink: line 2: exec: xcowsay: not found
```

**Comment by dasJ** [comment]
I could wrap the tools so they find xcowsay. But should I also wrap xcowfortune? I feel like the dependency to fortune isn't exactly necessary

**Comment by dasJ** [comment]
The maintainer updated to support GTK 3. I also wrapped the tools so it finds everything ;)

File: pkgs/games/xcowsay/default.nix [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  pname = "xcowsay";
+  version = "1.5";
+
+  src = fetchurl {
+    url = "http://www.nickg.me.uk/files/xcowsay-${version}.tar.gz";
+    sha256 = "0pyaa062z1ag26dhkm1yzp2hivnlmhlpqn5xg7mx9r1m652mm91y";
+  };
+
+  buildInputs = [ dbus dbus-glib gtk3 ];
+  nativeBuildInputs = [ makeWrapper pkg-config ];
```
Comment (jtojnar) [inline pkgs/games/xcowsay/default.nix]: We need `wrapGAppsHook` to be able to load images using `gdk-pixbuf`.

```ShellSession
$ env GDK_PIXBUF_MODULE_FILE= result/bin/xcowsay "foo"
Failed to load cow image: /nix/store/6qffdpj5j3dhz14bjhgcvnmmsnbm67av-xcowsay-1.5/share/xcowsay/cow_med.png
```

File: pkgs/games/xcowsay/default.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+stdenv.mkDerivation rec {
+  pname = "xcowsay";
+  version = "1.5";
+
+  src = fetchurl {
+    url = "http://www.nickg.me.uk/files/xcowsay-${version}.tar.gz";
+    sha256 = "0pyaa062z1ag26dhkm1yzp2hivnlmhlpqn5xg7mx9r1m652mm91y";
+  };
+
+  buildInputs = [ dbus dbus-glib gtk3 ];
```
Comment (jtojnar) [inline pkgs/games/xcowsay/default.nix]:
```suggestion
  buildInputs = [
    dbus-glib
    gtk3
    gdk-pixbuf # loading cow images
    librsvg # dreaming SVG images
  ];
```

**Comment by flokli** [comment]
Is there a `wlcowsay` too? :laughing:

**Comment by dasJ** [comment]
@GrahamcOfBorg build xcowsay

**Pull request merged**